2020/12/rain/src/main.rs

```diff
@@ -0,0 +1,440 @@
+use std::env;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+
+fn main() -> Result<(), Error> {
+    let args: Vec<String> = env::args().collect();
+    let mut reader = BufReader::new(File::open(&args[1])?);
+    let script = Action::parse_script_from_bufread(&mut reader)?;
+
+    let mut ship = Ship::new();
+    for action in script {
+        ship = ship.apply_part2(&action);
+    }
+
+    println!(
+        "The ship is at a manhattan distance of {}",
+        ship.manhattan_distance()
+    );
+
+    Ok(())
+}
+
+#[derive(Debug)]
+pub enum Error {
+    IO(std::io::Error),
+    Parse(std::num::ParseIntError),
+    TooShort,
+    WrongDirection(char),
+}
+
+impl From<std::io::Error> for Error {
+    fn from(ioe: std::io::Error) -> Self {
+        Error::IO(ioe)
+    }
+}
+
+impl From<std::num::ParseIntError> for Error {
+    fn from(pie: std::num::ParseIntError) -> Self {
+        Error::Parse(pie)
+    }
+}
+
+pub enum Action {
+    North(isize),
+    South(isize),
+    East(isize),
+    West(isize),
+    Left(isize),
+    Right(isize),
+    Forward(isize),
+}
+
+impl Action {
+    pub fn parse_script_from_string(script: &str) -> Result<Vec<Self>, Error> {
+        let mut res = Vec::new();
+        for line in script.lines() {
+            res.push(Action::parse(line)?);
+        }
+        Ok(res)
+    }
+
+    pub fn parse_script_from_bufread<T: BufRead>(script: &mut T) -> Result<Vec<Self>, Error> {
+        let mut res = Vec::new();
+        for line in script.lines() {
+            res.push(Action::parse(&line?)?);
+        }
+        Ok(res)
+    }
+
+    pub fn parse(input: &str) -> Result<Self, Error> {
+        let val = input[1..].parse::<isize>()?;
+        match input.chars().nth(0) {
```

Comment:
I'd just use `input[0]` here. You know you're dealing with single-byte characters. I think that produces a `u8` instead of a `char` to compare against in the match arms but you can use byte string literals:

```
'A' // type is char (u32)
b'A' // type is u8
"abc" // type is str ([u8] but checked to be UTF-8)
b"abc" // type is [u8]
```

Reply:
Oh, nice.  I'll keep that in mind.